rskj-core/src/main/java/co/rsk/net/sync/DownloadingBodiesSyncState.java

```diff
@@ -1,69 +1,301 @@
+        // we start from the last chunk that can be downloaded
+        while (segmentNumber >= 0){
+            Stack<Integer> chunks = chunksBySegment.get(segmentNumber);
+            // if the segment stack is empty then continue to next segment
+            if (!chunks.empty()) {
+                Integer chunkNumber = chunks.pop();
+                Stack<BlockHeader> headers = pendingHeaders.get(chunkNumber);
+                if (!headers.empty()) {
+                    chunksBeingDownloaded.put(peerId, chunkNumber);
+                    segmentsBeingDownloaded.put(peerId, segmentNumber);
+                    return Optional.of(headers.pop());
+                } else {
+                    // log something we are in trouble
+                }
+            }
+            segmentNumber--;
+        }
+        return Optional.empty();
     }
 
     @Override
     public void onEnter() {
-        requestBody();
+        initializeSegments();
+        startDownloading(suitablePeers);
+    }
+
+    private void startDownloading(List<NodeID> peers) {
+        for (int i = 0; i < peers.size(); i++){
+            NodeID peerId = peers.get(i);
+            Optional<BlockHeader> nextHeader = tryFindBlockHeader(peerId);
+            if (nextHeader.isPresent()){
+                requestBody(peerId, nextHeader.get());
+            }
+        }
     }
 
-    private void requestBody(){
-        BlockHeader header = pendingHeaders.remove();
-        long messageId = syncEventsHandler.sendBodyRequest(header);
-        NodeID peerId = syncInformation.getSelectedPeerId();
+    @Override
+    public void tick(Duration duration) {
+        List<NodeID> timeoutedNodes = timeElapsedByPeer.entrySet().stream()
+                .filter(e -> chunksBeingDownloaded.containsKey(e.getKey()) &&
+                        e.getValue().plus(duration).compareTo(syncConfiguration.getTimeoutWaitingRequest()) >= 0)
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toList());
+
+        for (NodeID peerId: timeoutedNodes){
+            syncInformation.reportEvent("Timeout waiting requests from node {}",
+                    EventType.TIMEOUT_MESSAGE, peerId);
+
+            Long messageId = messagesByPeers.remove(peerId);
+            BlockHeader header = pendingBodyResponses.remove(messageId).header;
+            clearPeerInfo(peerId);
+            resetChunkAndHeader(peerId, header);
+        }
+
+        if (suitablePeers.size() == 0){
+            syncEventsHandler.stopSyncing();
+            return;
+        }
+
+        startDownloading(getInactivePeers());
+
+        if (chunksBeingDownloaded.size() == 0){
+            syncEventsHandler.stopSyncing();
+        }
+    }
+
+    private List<NodeID> getInactivePeers() {
+        return suitablePeers.stream()
+                .filter(p -> !chunksBeingDownloaded.containsKey(p))
+                .collect(Collectors.toList());
+    }
+
+    private void initializeSegments() {
```

Comment:
This method is pretty hard to understand. I would add comments.

Reply:
Added some comments to explain motivation on segments and how this method should work.